nimbus/src/lib.rs

```diff
@@ -38,10 +41,13 @@ const DB_KEY_NIMBUS_ID: &str = "nimbus-id";
 /// It should hold all the information needed to communicate a specific user's
 /// experimentation status
 pub struct NimbusClient {
-    settings_client: Box<dyn SettingsClient + Send>,
+    settings_client: Mutex<Box<dyn SettingsClient + Send>>,
     available_randomization_units: AvailableRandomizationUnits,
     app_context: AppContext,
-    db: OnceCell<Database>,
+    db: OnceCell<Mutex<Database>>,
+    // An in-memory cache of our enrollments so that we can answer
+    // get_experiment_branch without doing (or waiting for) IO.
+    cached_experiment_branches: RwLock<RefCell<Option<HashMap<String, String>>>>,
```

Comment:
code-organization nano-nit: you could consider pulling this four-levels-of-nesting-deep type out into a little wrapper `CachedExperimentBranches` struct, along with some helper methods for its core update logic. This would keep the methods of the main `NimbusClient` struct more readable and might give you an obvious place to document the design choices behind it.

Not a big deal though.

Reply:
Good idea - I'll have a poke at that next week.

Reply:
I came up with something similar to what James suggests.

> when we invariably need to expose more methods to the main thread. e.g. `get_all_experiments()`, `get_global_user_participation()`, `get_experiment_branches()`.

While I agree that is inevitable there will be others, we should resist doing so where possible - ie, we should make sure there *really* is a requirement for doing this from the main thread to avoid creeping complexity. For example, I doubt that any of the functions you mention will need that in practice as they are most likely to be used by the app UI and not directly by experiments (but again, agree that there might well be others)